File: library/src/main/java/com/senierr/adapter/internal/ViewHolderWrapper.java
package com.senierr.adapter.internal;

import android.support.annotation.IdRes;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.widget.RecyclerView;
import android.util.SparseArray;
import android.view.View;
import android.view.ViewGroup;

import com.senierr.adapter.listener.OnChildClickListener;
import com.senierr.adapter.listener.OnChildLongClickListener;
import com.senierr.adapter.listener.OnItemClickListener;
import com.senierr.adapter.listener.OnItemLongClickListener;

import java.util.List;

/**
 * 视图-数据封装
 *
 * @author zhouchunjie
 * @date 2017/9/25
 */
public abstract class ViewHolderWrapper<T> {

    private @Nullable MultiTypeAdapter multiTypeAdapter;
    private @Nullable OnItemClickListener<T> onItemClickListener;
    private @Nullable OnItemLongClickListener<T> onItemLongClickListener;
    private @Nullable SparseArray<OnChildClickListener<T>> onChildClickListeners;
    private @Nullable SparseArray<OnChildLongClickListener<T>> onChildLongClickListeners;

    @NonNull
    final ViewHolder generateViewHolder(@NonNull ViewGroup parent,
                                        @NonNull MultiTypeAdapter multiTypeAdapter) {
        setMultiTypeAdapter(multiTypeAdapter);
        // 创建视图
        final ViewHolder holder = onCreateViewHolder(parent);
        // 绑定列表项点击事件
        if (onItemClickListener != null) {
            holder.itemView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    int position = holder.getLayoutPosition();
                    T t = getItemData(position);
                    if (t != null) {
                        onItemClickListener.onClick(holder, position, t);
                    }
                }
            });
        }
        // 绑定列表项长按事件
        if (onItemLongClickListener != null) {
            holder.itemView.setOnLongClickListener(new View.OnLongClickListener() {
                @Override
                public boolean onLongClick(View view) {
                    int position = holder.getLayoutPosition();
                    T t = getItemData(position);
                    return t != null && onItemLongClickListener.onLongClick(holder, position, t);
                }
            });
        }
        // 绑定子视图点击事件
        if (onChildClickListeners != null && onChildClickListeners.size() > 0) {
            for (int i = 0; i < onChildClickListeners.size(); i++) {
                int key = onChildClickListeners.keyAt(i);
                final OnChildClickListener<T> onChildClickListener = onChildClickListeners.get(key);
                View childView = holder.findView(key);
                if (onChildClickListener != null && childView != null) {
                    childView.setOnClickListener(new View.OnClickListener() {
                        @Override
                        public void onClick(View view) {
                            int position = holder.getLayoutPosition();
                            T t = getItemData(position);
                            if (t != null) {
                                onChildClickListener.onClick(holder, view, position, t);
                            }
                        }
                    });
                }
            }
        }
        // 绑定子视图长按事件
        if (onChildLongClickListeners != null && onChildLongClickListeners.size() > 0) {
            for (int i = 0; i < onChildLongClickListeners.size(); i++) {
                int key = onChildLongClickListeners.keyAt(i);
                final OnChildLongClickListener<T> onChildLongClickListener = onChildLongClickListeners.get(key);
                View childView = holder.findView(key);
                if (onChildLongClickListener != null && childView != null) {
                    childView.setOnLongClickListener(new View.OnLongClickListener() {
                        @Override
                        public boolean onLongClick(View v) {
                            int position = holder.getLayoutPosition();
                            T t = getItemData(position);
                            return t != null && onChildLongClickListener.onLongClick(holder, v, position, t);
                        }
                    });
                }
            }
        }
        return holder;
    }

    final void setMultiTypeAdapter(@NonNull MultiTypeAdapter multiTypeAdapter) {
        this.multiTypeAdapter = multiTypeAdapter;
    }

    @Nullable
    @SuppressWarnings("unchecked")
    private T getItemData(int position) {
        if (multiTypeAdapter == null) return null;
        try {
            return (T) multiTypeAdapter.getData().get(position);
        } catch (Exception e) {
            return null;
        }
    }

    @Nullable
    public final MultiTypeAdapter getMultiTypeAdapter() {
        return multiTypeAdapter;
    }

    public long getItemId(T item) {
        return RecyclerView.NO_ID;
    }

    public void onViewRecycled(ViewHolder holder) {}

    public boolean onFailedToRecycleView(ViewHolder holder) {
        return false;
    }

    public void onViewAttachedToWindow(ViewHolder holder) {}

    public void onViewDetachedFromWindow(ViewHolder holder) {}

    public void onAttachedToRecyclerView(RecyclerView recyclerView) {}

    public void onDetachedFromRecyclerView(RecyclerView recyclerView) {}

    @NonNull
    public abstract ViewHolder onCreateViewHolder(@NonNull ViewGroup parent);

    public abstract void onBindViewHolder(@NonNull ViewHolder holder, @NonNull T item);

    public void onBindViewHolder(@NonNull ViewHolder holder, @NonNull T item, @NonNull List<Object> payloads) {
        onBindViewHolder(holder, item);
    }

    /**
     * 获取当前项所占列数
     *
     * @param item 数据项
     * @return 默认为1，超过最大列数为全行
     */
    public int getSpanSize(T item) {
        return 1;
    }

    @Nullable
    public final OnItemClickListener<T> getOnItemClickListener() {
        return onItemClickListener;
    }

    public void setOnItemClickListener(@Nullable OnItemClickListener<T> onItemClickListener) {
        this.onItemClickListener = onItemClickListener;
    }

    @Nullable
    public final OnItemLongClickListener<T> getOnItemLongClickListener() {
        return onItemLongClickListener;
    }

    public final void setOnItemLongClickListener(@Nullable OnItemLongClickListener<T> onItemLongClickListener) {
        this.onItemLongClickListener = onItemLongClickListener;
    }

    @Nullable
    public final OnChildClickListener<T> getOnChildClickListener(@IdRes int childId) {
        if (onChildClickListeners == null) return null;
        return onChildClickListeners.get(childId);
    }

    public final void setOnChildClickListener(@IdRes int childId, @Nullable OnChildClickListener<T> onChildClickListener) {
        if (onChildClickListeners == null) {
            onChildClickListeners = new SparseArray<>();
        }
        onChildClickListeners.put(childId, onChildClickListener);
    }

    @Nullable
    public final OnChildLongClickListener<T> getOnChildLongClickListener(@IdRes int childId) {
        if (onChildLongClickListeners == null) return null;
        return onChildLongClickListeners.get(childId);
    }

    public final void setOnChildLongClickListener(@IdRes int childId, @Nullable OnChildLongClickListener<T> onChildLongClickListener) {
        if (onChildLongClickListeners == null) {
            onChildLongClickListeners = new SparseArray<>();
        }
        onChildLongClickListeners.put(childId, onChildLongClickListener);
    }
}

File: multitype/src/main/java/com/senierr/adapter/internal/ViewHolderWrapper.kt
package com.senierr.adapter.internal

import androidx.annotation.IdRes
import androidx.annotation.LayoutRes
import androidx.recyclerview.widget.RecyclerView
import android.util.Log
import android.util.SparseArray
import android.view.View
import android.view.ViewGroup

/**
 * 视图-数据封装
 *
 * @author zhouchunjie
 * @date 2017/9/25
 */
abstract class ViewHolderWrapper<T>(@LayoutRes private val layoutId: Int = -1) {

    protected var recyclerView: RecyclerView? = null
    protected lateinit var multiTypeAdapter: MultiTypeAdapter

    private var onItemClickListener: ((viewHolder: ViewHolder, position: Int, item: T) -> Unit)? = null
    private var onItemLongClickListener: ((viewHolder: ViewHolder, position: Int, item: T) -> Boolean)? = null

    private var onChildClickListeners = SparseArray<(viewHolder: ViewHolder, view: View, position: Int, item: T) -> Unit>()
    private var onChildLongClickListeners = SparseArray<(viewHolder: ViewHolder, view: View, position: Int, item: T) -> Boolean>()

    /**
     * 注册回调
     */
    open fun onRegister(multiTypeAdapter: MultiTypeAdapter) {
        this.multiTypeAdapter = multiTypeAdapter
    }

    /**
     * 创建视图
     */
    internal fun onCreateViewHolderInternal(parent: ViewGroup): ViewHolder {
        val holder = onCreateViewHolder(parent)
        registerListeners(holder)
        return holder
    }

    /**
     * 创建视图
     */
    open fun onCreateViewHolder(parent: ViewGroup): ViewHolder {
        return ViewHolder.create(parent, layoutId)
    }

    open fun onViewAttachedToWindow(holder: ViewHolder) {}

    open fun onViewDetachedFromWindow(holder: ViewHolder) {}

    open fun onAttachedToRecyclerView(recyclerView: RecyclerView) {
        this.recyclerView = recyclerView
    }

    open fun onDetachedFromRecyclerView(recyclerView: RecyclerView) {
        this.recyclerView = null
    }

    abstract fun onBindViewHolder(holder: ViewHolder, item: T)

    open fun onBindViewHolder(holder: ViewHolder, item: T, payloads: List<Any>) {
        onBindViewHolder(holder, item)
    }

    /**
     * 获取当前项所占列数
     *
     * @param item 数据项
     * @return 默认为1，超过最大列数为全行
     */
    open fun getSpanSize(item: T): Int = 1

    @Suppress("UNCHECKED_CAST")
    private fun getItemData(position: Int): T? {
        var result: T? = null
        try {
            result = multiTypeAdapter.data[position] as T
        } catch (e: Exception) {
            Log.w(this.javaClass.simpleName, "The value in the position does not exist.")
        }
        return result
    }

    /**
     * 注册点击事件
     */
    private fun registerListeners(holder: ViewHolder) {
        // 绑定列表项点击事件
        onItemClickListener?.let { listener ->
            holder.itemView.setOnClickListener {
                val position = holder.layoutPosition
                val t = getItemData(position)
                if (t != null) listener.invoke(holder, position, t)
            }
        }

        // 绑定列表项长按事件
        onItemLongClickListener?.let { listener ->
            holder.itemView.setOnLongClickListener {
                val position = holder.layoutPosition
                val t = getItemData(position)
                t != null && listener.invoke(holder, position, t)
            }
        }

        // 绑定子视图点击事件
        if (onChildClickListeners.size() > 0) {
            for (i in 0 until onChildClickListeners.size()) {
                val key = onChildClickListeners.keyAt(i)
                val onChildClickListener = onChildClickListeners.get(key)
                val childView = holder.findView<View>(key)
                if (onChildClickListener != null) {
                    childView?.setOnClickListener { view ->
                        val position = holder.layoutPosition
                        val t = getItemData(position)
                        if (t != null) onChildClickListener.invoke(holder, view, position, t)
                    }
                }
            }
        }

        // 绑定子视图长按事件
        if (onChildLongClickListeners.size() > 0) {
            for (i in 0 until onChildLongClickListeners.size()) {
                val key = onChildLongClickListeners.keyAt(i)
                val onChildLongClickListener = onChildLongClickListeners.get(key)
                val childView = holder.findView<View>(key)
                if (onChildLongClickListener != null) {
                    childView?.setOnLongClickListener { view ->
                        val position = holder.layoutPosition
                        val t = getItemData(position)
                        t != null && onChildLongClickListener.invoke(holder, view, position, t)
                    }
                }
            }
        }
    }

    fun setOnItemClickListener(listener: (viewHolder: ViewHolder, position: Int, item: T) -> Unit) {
        onItemClickListener = listener
    }

    fun setOnItemLongClickListener(listener: (viewHolder: ViewHolder, position: Int, item: T) -> Boolean) {
        onItemLongClickListener = listener
    }

    fun setOnChildClickListener(@IdRes childId: Int, listener: (viewHolder: ViewHolder, view: View, position: Int, item: T) -> Unit) {
        onChildClickListeners.put(childId, listener)
    }

    fun setOnChildLongClickListener(@IdRes childId: Int, listener: (viewHolder: ViewHolder, view: View, position: Int, item: T) -> Boolean) {
        onChildLongClickListeners.put(childId, listener)
    }
}